UI: Detect cpp memory leaks

cg2121 opened the pull request:
### Description
This overloads the cpp new/delete operators, to add an alloc counter
to detect memory leaks. This number is printed when using OBS in debug
mode when exiting.

### Motivation and Context
Makes it easier to find memory leaks in QT or cpp.

### How Has This Been Tested?
Ran the program and checked if the memory leak number was printed when exiting.

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

notr1ch commented on UI/obs-app.cpp:
```diff
@@ -69,6 +69,25 @@
 using namespace std;
+long alloc_count;
+
+void *operator new(size_t size)
+{
+	alloc_count++;
+	return malloc(size);
+}
```
This violates `new` semantics, if allocation fails a `std::bad_alloc` must be thrown.

cg2121 replied:
Added the std::bad_alloc exception.

notr1ch reviewed (commented):
I'm generally not a fan of big changes to global program state like this for such a small feature. Perhaps it could be enabled in debug only builds, but there are far better tools for finding memory leaks, and such tools may expect non-overloaded `new` / `delete` implementations in order to function.

e00E commented:
This approach is weird to me a for a number of reasons:

1. Will replacing new and delete like this have side effects like different behavior compared to the default versions? For example it is unclear whether the default implementation calls malloc.
2. This implementation is not thread safe. Do these functions need to be threadsafe?
3. When you log the counter you are at the end of main. But at this point destructors from local objects have not yet been called yet so you can get false positives.

A much more complete approach to this is compiler instrumentation like https://clang.llvm.org/docs/LeakSanitizer.html . I do not see the advantage of this implementation over that.

cg2121 commented:
Updated to only overload new/delete in debug mode. Also updated to use bzalloc/bfree to use the internal OBS alloc counter, and it seems the program still works as expected.

cg2121 commented:
Closing, as explained, there are other ways to do this.
